### PR TITLE
[WIP] MAT Layers: Embedding + Generator

### PR DESCRIPTION
**Pull request by atreyamaj**
## Description
This PR contains the Embedding and Generator layers for MAT.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -47,3 +48,144 @@ def __init__(self, scale, eps=1e-5):
...
+  .. [1] Lukasz Maziarka et al. "Molecule Attention Transformer" Graph Representation Learning workshop and Machine Learning and the Physical Sciences workshop at NeurIPS 2019. 2020. https://arxiv.org/abs/2002.08264
+
+  Examples
+  --------
+  >>> import deepchem as dc
+  >>> layer = MATGenerator(hsize = 1024, aggregation_type = 'mean', d_output = 1, n_layers = 1, attn_hidden = 128, attn_out = 4)
+  """
+
+  def __init__(self,
+               *,
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:114]: Same * and missing types problem again?

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -47,3 +48,144 @@ def __init__(self, scale, eps=1e-5):
...
+    d_output: int
+      Size of output layer.
+    dropout_p: float
+      Dropout probability for layer.
+    attn_hidden: int
+      Size of hidden attention layer.
+    attn_out: int
+      Size of output attention layer.
+    """
+
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:139]: Maybe this space can go. But if yapf doesn't complain, I guess its fine.

File: deepchem/models/torch_models/layers.py [hunk 3]
```diff
@@ -14,19 +15,19 @@ class ScaleNorm(nn.Module):
...
   Examples
   --------
   >>> import deepchem as dc
   >>> scale = 0.35
-  >>> layer = dc.models.torch_models.ScaleNorm(scale)
-  >>> input_tensor = torch.tensor([[1.269, 39.36], [0.00918, -9.12]])
-  >>> output_tensor = layer.forward(input_tensor)
+  >>> layer = dc.models.torch_models.layers.ScaleNorm(scale)
+  >>> input_tensor = torch.Tensor([[1.269, 39.36], [0.00918, -9.12]])
```
[the remark this answers is not in the document]
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py:29]: While torch.Tensor(array) seems to be working too, I have changed it to torch.tensor now.

File: deepchem/models/torch_models/layers.py [hunk 4]
```diff
@@ -47,3 +48,144 @@ def __init__(self, scale, eps=1e-5):
...
+    d_input: int
+      Size of input layer.
+    d_output: int
+      Size of output layer.
+    dropout_p: float
+      Dropout probability for layer.
+    """
+
+    super(MATEmbedding, self).__init__()
+    self.lut = nn.Linear(d_input, d_output)
```
[the remark this answers is not in the document]
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py:83]: It's basically a linear unit. I have renamed it now.

File: deepchem/models/torch_models/layers.py [hunk 5]
```diff
@@ -47,3 +48,144 @@ def __init__(self, scale, eps=1e-5):
...
+  ----------
+  .. [1] Lukasz Maziarka et al. "Molecule Attention Transformer" Graph Representation Learning workshop and Machine Learning and the Physical Sciences workshop at NeurIPS 2019. 2020. https://arxiv.org/abs/2002.08264
+
+  Examples
+  --------
+  >>> import deepchem as dc
+  >>> layer = MATEmbedding(d_input = 1024, d_output = 1024, dropout_p = 0.2)
+  """
+
+  def __init__(self, *, d_input: int, d_output: int, dropout_p: float):
```
[the remark this answers is not in the document]
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py]: Yes, I have removed it now, thank you for the heads up!

File: deepchem/models/torch_models/layers.py [hunk 6]
```diff
@@ -14,36 +15,175 @@ class ScaleNorm(nn.Module):
...
   >>> scale = 0.35
-  >>> layer = dc.models.torch_models.ScaleNorm(scale)
+  >>> layer = dc.models.torch_models.layers.ScaleNorm(scale)
   >>> input_tensor = torch.tensor([[1.269, 39.36], [0.00918, -9.12]])
-  >>> output_tensor = layer.forward(input_tensor)
+  >>> output_tensor = layer(input_tensor)
   """
 
-  def __init__(self, scale, eps=1e-5):
+  def __init__(self, scale: int, eps: float = 1e-5):
```
Comment (vsomnath) [inline deepchem/models/torch_models/layers.py:33]: int or float?

File: deepchem/models/torch_models/layers.py [hunk 7]
```diff
@@ -14,36 +15,175 @@ class ScaleNorm(nn.Module):
...
     """Initialize a ScaleNorm layer.
 
     Parameters
     ----------
     scale: Real number or single element tensor
       Scale magnitude.
     eps: float
-      Epsilon value.
+      Epsilon value. Default = 1e-5.
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:41]: We usually do something like:

```
eps: float (default 1e-5)
  Epsilon value
```

File: deepchem/models/torch_models/layers.py [hunk 8]
```diff
@@ -14,36 +15,175 @@ class ScaleNorm(nn.Module):
...
+  In the MATEmbedding layer, an input tensor is processed through a dropout-adjusted linear layer and the resultant vector is returned.
+
+  References
+  ----------
+  .. [1] Lukasz Maziarka et al. "Molecule Attention Transformer" Graph Representation Learning workshop and Machine Learning and the Physical Sciences workshop at NeurIPS 2019. 2020. https://arxiv.org/abs/2002.08264
+
+  Examples
+  --------
+  >>> import deepchem as dc
+  >>> layer = MATEmbedding(d_input = 1024, d_output = 1024, dropout_p = 0.2)
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:65]: Can you show a usage example of how to invoke the layer on sample input?

File: deepchem/models/torch_models/layers.py [hunk 9]
```diff
@@ -14,36 +15,175 @@ class ScaleNorm(nn.Module):
...
+  Depending on the type of aggregation selected, the attention output layer performs different operations.
+
+  References
+  ----------
+  .. [1] Lukasz Maziarka et al. "Molecule Attention Transformer" Graph Representation Learning workshop and Machine Learning and the Physical Sciences workshop at NeurIPS 2019. 2020. https://arxiv.org/abs/2002.08264
+
+  Examples
+  --------
+  >>> import deepchem as dc
+  >>> layer = MATGenerator(hsize = 1024, aggregation_type = 'mean', d_output = 1, n_layers = 1, attn_hidden = 128, attn_out = 4)
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:108]: Can you add an example showing how to invoke the layer on input?

File: deepchem/models/torch_models/layers.py [hunk 10]
```diff
@@ -2,6 +2,7 @@
 try:
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:2]: Can you add unit tests for the new layers?

**Review by rbharath — COMMENTED** [review]
Did a first pass. We still need unit tests before we can merge this in :)

File: deepchem/models/torch_models/layers.py (hunk 7 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py:41]: Changed! I've added this change to this branch (`GeneratorPR`)

File: deepchem/models/torch_models/layers.py (hunk 8 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py:65]: Added.

File: deepchem/models/torch_models/layers.py (hunk 9 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py]: Added!

File: deepchem/models/torch_models/layers.py (hunk 10 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py:2]: Working on it :+1:

**Review by rbharath — APPROVED** [review]
CI looks fine and PR looks good to going to go ahead and merge in

**Pull request merged**